controllers/users.js

```diff
@@ -0,0 +1,410 @@
+  }
+};
+
+var getAllUserSettings = function(callback) {
+  var opts = {
+    include_docs: true,
+    key: ['user-settings']
+  };
+  db.medic.view('medic', 'doc_by_type', opts, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var getAllUsers = function(callback) {
+  db._users.list({include_docs: true}, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, results.rows);
+  });
+};
+
+var getFacilities = function(callback) {
+  db.medic.view('medic', 'facilities', {include_docs: true}, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var isAPlace = function(place) {
+  return [
+    'district_hospital',
+    'health_center',
+    'clinic'
+  ].indexOf(place.type) !== -1;
+};
+
+var getPlace = function(id, callback) {
+  db.medic.get(id, function(err, place) {
+    if (err) {
+      return callback(err);
+    }
+    if (!isAPlace(place)) {
+      return callback('Wrong type, this is not a place.');
+    }
+    callback(null, place);
+  });
+};
+
+var createUser = function(data, response, callback) {
+  response = response || {};
+  var id = createID(data.username),
+      user = getUserUpdates(id, data);
+  db._users.insert(user, id, function(err, body) {
+    if (err) {
+      return callback('Failed to create user. ' + err);
+    }
+    response.user = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+/*
+ * Warning: not doing validation of the contact data against a form yet.  The
+ * form is user defined in settings so being liberal with what gets saved to
+ * the database. Ideally CouchDB could validate a given object against a form
+ * in validate_doc_update.
+ */
+var createContact = function(data, response, callback) {
+  response = response || {};
+  db.medic.insert(data.contact, function(err, body) {
+    if (err) {
+      return callback('Failed to create contact. ' + err);
+    }
+    // save contact id for user settings
+    data.contact = body.id;
+    response.contact = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+var createUserSettings = function(data, response, callback) {
+  response = response || {};
+  var settings = getSettingsUpdates(data);
+  db.medic.insert(settings, createID(data.username), function(err, body) {
+    if (err) {
+      return callback('Failed to create user settings. ' + err);
+    }
+    response['user-settings'] = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+var hasParent = function(facility, id) {
+  if (facility._id === id) {
+    return true;
+  }
+  // clone facility before modifying it
+  var f = JSON.parse(JSON.stringify(facility));
```

Comment:
I think we can probably avoid writing to `facility` here, which would remove the need for the clone AFAICT.